src/drive/web/modules/viewer/FilesViewerV2.jsx

```diff
@@ -0,0 +1,153 @@
+  // a direct link to a file that wasn't in the first 50 files of the containing folder
+  // (it comes from a fetchMore...) ; we load the file attributes directly as a contingency measure
+  async fetchFileIfNecessary() {
+    if (this.getCurrentIndex() !== -1) return
+    if (this.state.currentFile && this._mounted) {
+      this.setState({ currentFile: null })
+    }
+
+    const { fileId, client } = this.props
+    try {
+      const { data } = await client.query(client.get('io.cozy.files', fileId))
+      this.setState({
+        currentFile: data
+      })
+    } catch (e) {
+      logger.warn("can't find the file")
+      this.onClose()
+    }
+  }
+
+  fetchMoreIfNecessary() {
+    const { files, fileId, filesQuery } = this.props
+    const fileCount = filesQuery.count
+    // If we get close of the last file fetched, but we know there are more in the folder
+    // (it shouldn't happen in /recent), we fetch more files
+    const currentIndex = files.findIndex(f => f.id === fileId)
+
+    if (files.length !== fileCount && files.length - currentIndex <= 5) {
+      filesQuery.fetchMore()
+    }
+  }
+
+  onClose = () => {
+    const { router, folderId } = this.props
+    router.push({
+      pathname: getFolderPath(folderId)
+    })
+  }
+
+  onChange = nextFile => {
+    const { router, folderId } = this.props
+    router.push({
+      pathname: getViewerPath(folderId, nextFile.id)
+    })
+  }
+
+  getCurrentIndex() {
+    const { files, fileId } = this.props
+    return files.findIndex(f => f.id === fileId)
+  }
+
+  render() {
+    const { t, files } = this.props
+    const currentIndex = this.getCurrentIndex(files)
+
+    // If we can't find the file, we fallback to the (potentially loading)
+    // direct stat made by the viewer
+    if (currentIndex === -1 && !this.state.currentFile) {
+      return <FilesViewerLoading />
+    } else {
+      const hasCurrentIndex = currentIndex != -1
+      const viewerFiles = hasCurrentIndex ? files : [this.state.currentFile]
+      const viewerIndex = hasCurrentIndex ? currentIndex : 0
+
+      return (
+        <Overlay>
+          <Viewer
+            files={viewerFiles}
+            currentIndex={viewerIndex}
+            onChangeRequest={this.onChange}
+            onCloseRequest={this.onClose}
+            renderFallbackExtraContent={file => <Fallback file={file} t={t} />}
+          />
+        </Overlay>
+      )
+    }
+  }
+}
+
+const getViewableFiles = files => {
+  return files.filter(f => f.type !== 'directory')
+}
+
+const mapStateToProps = state => ({
+  folderId: getCurrentFolderId(state),
+  fileId: getCurrentFileId(state)
+})
+
+const FilesViewerWithQuery = ({ ...props }) => {
+  const filesQuery = buildQuery({
+    currentFolderId: props.folderId,
+    type: 'file',
+    sortAttribute: 'name',
+    sortOrder: 'asc'
+  })
+  const results = useQuery(filesQuery.definition, filesQuery.options)
+  if (results.data) {
+    const viewableFiles = getViewableFiles(results.data)
```

Comment:
I think a neat side effect o this change is that we can remove `getViewableFiles`, as we filter by type in the query!